pkg/frontend/querymiddleware/spin_off_subqueries.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	return m
+}
+
+func newSpinOffSubqueriesMiddleware(
+	limits Limits,
+	logger log.Logger,
+	engine *promql.Engine,
+	rangeHandler MetricsQueryHandler,
+	registerer prometheus.Registerer,
+	defaultStepFunc func(int64) int64,
+) MetricsQueryMiddleware {
+	metrics := newSpinOffSubqueriesMetrics(registerer)
+
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &spinOffSubqueriesMiddleware{
+			next:            next,
+			limits:          limits,
+			logger:          logger,
+			engine:          engine,
+			rangeHandler:    rangeHandler,
+			metrics:         metrics,
+			defaultStepFunc: defaultStepFunc,
+		}
+	})
+}
+
+func (s *spinOffSubqueriesMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	// Log the instant query and its timestamp in every error log, so that we have more information for debugging failures.
+	logger := log.With(s.logger, "query", req.GetQuery(), "query_timestamp", req.GetStart())
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, logger, "spinOffSubqueriesMiddleware.Do")
+	defer spanLog.Span.Finish()
+
+	// For now, the feature is completely opt-in
+	// So we check that the given query is allowed to be spun off
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	matched := false
+	for _, tenantID := range tenantIDs {
+		patterns := s.limits.InstantQueriesWithSubquerySpinOff(tenantID)
+
+		for _, pattern := range patterns {
+			matcher, err := labels.NewFastRegexMatcher(pattern)
+			if err != nil {
+				level.Error(spanLog).Log("msg", "failed to compile regex pattern", "pattern", pattern, "err", err)
+				continue
+			}
+
+			if matcher.MatchString(req.GetQuery()) {
+				matched = true
+				break
+			}
+		}
+
+		if matched {
+			break
+		}
+	}
+
+	if !matched {
+		spanLog.DebugLog("msg", "subquery spin off is disabled for this query")
+		return s.next.Do(ctx, req)
+	}
+
+	// Increment total number of instant queries attempted to spin off subqueries from.
+	s.metrics.spinOffAttempts.Inc()
+
+	mapperStats := astmapper.NewSubquerySpinOffMapperStats()
+	mapperCtx, cancel := context.WithTimeout(ctx, shardingTimeout)
+	defer cancel()
+	mapper := astmapper.NewSubquerySpinOffMapper(mapperCtx, s.defaultStepFunc, spanLog, mapperStats)
+
+	expr, err := parser.ParseExpr(req.GetQuery())
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to parse query", "err", err)
+		s.metrics.spinOffSkipped.WithLabelValues(subquerySpinoffSkippedReasonParsingFailed).Inc()
+		return nil, apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	}
+
+	spinOffQuery, err := mapper.Map(expr)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) && ctx.Err() == nil {
+			level.Error(spanLog).Log("msg", "timeout while spinning off subqueries, please fill in a bug report with this query, falling back to try executing without spin-off", "err", err)
```

Comment:
[nit] In some places we use "spin off" and in others we use "spin-off" - would be good to be consistent about this